3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,271 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {dev, user} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../src/3p-analytics-common';
+initLogConstructor();
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    let frameNameData = null;
+    try {
+      frameNameData = JSON.parse(this.win_.name);
+      if (!frameNameData.sentinel) {
+        throw new Error('JSON parsed, but did not include sentinel.');
+      }
+      // The sentinel is required by iframe-messaging-client, and is used to
+      // uniquely identify the frame as part of message routing
+      /** @const {string} */
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel from name attr: ' +
+        this.win_.name + ': ' + e.message);
+      return;
+    }
+    this.sentinel_ = frameNameData && frameNameData.sentinel;
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding AmpAnalytics3pCreativeMessageRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+      messageContainer => {
+        dev().assert(messageContainer.data,
```

Comment:
Done (Object.entries(messageContainer.data).length since it's an object not an array)